core/services/offchainreporting/oracle.go

```diff
@@ -0,0 +1,198 @@
+package offchainreporting
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/jinzhu/gorm"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-peerstore/pstoremem"
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/store/offchainreportingdb"
+	"github.com/smartcontractkit/chainlink/core/store/orm"
+	"github.com/smartcontractkit/chainlink/core/utils"
+	ocrnetworking "github.com/smartcontractkit/offchain-reporting/lib/networking"
+	ocr "github.com/smartcontractkit/offchain-reporting/lib/offchainreporting"
+	ocrtypes "github.com/smartcontractkit/offchain-reporting/lib/offchainreporting/types"
+	ocrcontracts "github.com/smartcontractkit/offchain-reporting/lib/prototype/contracts"
+)
+
+const JobType job.Type = "offchainreporting"
+
+func RegisterJobType(
+	db *gorm.DB,
+	config *orm.Config,
+	keyStore *KeyStore,
+	jobSpawner job.Spawner,
+	pipelineRunner pipeline.Runner,
+	ethClient eth.Client,
+	logBroadcaster eth.LogBroadcaster,
+) {
+	jobSpawner.RegisterDelegate(
+		NewJobSpawnerDelegate(db, config, keyStore, pipelineRunner, ethClient, logBroadcaster),
+	)
+}
+
+type jobSpawnerDelegate struct {
+	db             *gorm.DB
+	config         *orm.Config
+	keyStore       *KeyStore
+	pipelineRunner pipeline.Runner
+	ethClient      eth.Client
+	logBroadcaster eth.LogBroadcaster
+}
+
+func NewJobSpawnerDelegate(
+	db *gorm.DB,
+	config *orm.Config,
+	keyStore *KeyStore,
+	pipelineRunner pipeline.Runner,
+	ethClient eth.Client,
+	logBroadcaster eth.LogBroadcaster,
+) *jobSpawnerDelegate {
+	return &jobSpawnerDelegate{db, config, keyStore, pipelineRunner, ethClient, logBroadcaster}
+}
+
+func (d jobSpawnerDelegate) JobType() job.Type {
+	return JobType
+}
+
+func (d jobSpawnerDelegate) ToDBRow(spec job.Spec) models.JobSpecV2 {
+	concreteSpec, ok := spec.(OracleSpec)
+	if !ok {
+		panic(fmt.Sprintf("expected an offchainreporting.OracleSpec, got %T", spec))
+	}
+	return models.JobSpecV2{OffchainreportingOracleSpec: &concreteSpec.OffchainReportingOracleSpec}
+}
+
+func (d jobSpawnerDelegate) FromDBRow(spec models.JobSpecV2) job.Spec {
+	if spec.OffchainreportingOracleSpec == nil {
+		return nil
+	}
+	return &OracleSpec{
+		OffchainReportingOracleSpec: *spec.OffchainreportingOracleSpec,
+		jobID:                       spec.ID,
+	}
+}
+
+func (d jobSpawnerDelegate) ServicesForSpec(spec job.Spec) ([]job.Service, error) {
+	concreteSpec := spec.(*OracleSpec)
+
+	aggregator, err := ocrcontracts.NewOffchainReportingAggregator(
+		concreteSpec.ContractAddress,
+		d.ethClient,
+		d.logBroadcaster,
+		concreteSpec.JobID(),
+		nil, // auth *bind.TransactOpts,
```

Comment:
Just put a TODO here since it changes considerably once https://github.com/smartcontractkit/chainlink/pull/3407/files goes in